### ISSUE TITLE
Check returns to be compatible with declaration

### ISSUE THREAD

**Issue by kelunik** [kalessil/phpinspectionsea#668]
Got a bug report in https://github.com/amphp/beanstalk/issues/11. Thought EA should probably have something that checks it if PhpStorm doesn't have it by default, but it turns out there's nothing.

Return values should be checked to be compatible to the declared return type.

**Example**

```php
    public function delete(int $id): Promise {
        $payload = "delete $id\r\n";

        return $this->send($payload, function (array $response): int {
            list($type) = $response;

            switch ($type) {
                case "DELETED":
                    return true;

                case "NOT_FOUND":
                    return false;

                default:
                    throw new BeanstalkException("Unknown response: " . $type);
            }
        });
    }
```

**Expected Result**

`return true;` and `return false;` should be flagged as probable bugs, even if the code "works fine" with weak types.

**Comment by funivan** [comment]
Do you have statement `declare(strict_types = 1);` at the top of the file?

![08 46 22](https://user-images.githubusercontent.com/425208/33393061-91722c34-d546-11e7-9660-fce18ee43b05.png)
All wokrs fine for me.

This inspection is from PS: `Strict type checking rules violation`

**Comment by kelunik** [comment]
No, that's why it should be marked as probable bug even if it works.

**Comment by funivan** [comment]
Hm. When values can be automatically converted then we do not have any problems:
```php
<?php
  $fn = function () : int{     return true;   };
  var_dump($fn());
```
Without declare strict all works fine: you get `int(1)`

In another example 
```php
$fn = function () : int{
    return new stdClass();
  };
  var_dump($fn());
```
You will get type error. I thnik this is a case: **when we have types that can`t be conveted**

@kalessil do you want to take it or it will be better to report this issue to the PS team?

**Comment by kalessil** [comment]
If I understand correctly, following is expected:
- if return type has been declared, and `declare(strict_types = 1);` is not declared, behave as `Strict type checking rules violation`

If I remember correctly PS team has declined or didn't proceed further with similar FRs.
I'd check if we can implement this in Extended-version.

**Comment by orklah** [comment]
> If I understand correctly, following is expected:
> 
> * if return type has been declared, and `declare(strict_types = 1);` is not declared, behave as `Strict type checking rules violation`
> 
> If I remember correctly PS team has declined or didn't proceed further with similar FRs.
> I'd check if we can implement this in Extended-version.

PHPStorm introduced a new setting in the "Strict type checking rules violation" inspection in EAP to activate the check even if declare(strict_types = 1); is missing in the file.

The code in example now triggers an error when this setting is activated. I think we may close this issue.

**Comment by funivan** [comment]
I just check the latest inspection from the PHPStorm team `Strict type checking rules violation` and it works great. 

@kalessil you can close this issue. JB team make a great work)
